python/dgl/dataloading/dataloader.py

```diff
@@ -248,33 +253,8 @@ def sample_blocks(self, g, seed_nodes, exclude_eids=None):
                     for ntype, nodes in seed_nodes_in.items()}
             else:
                 seed_nodes_in = seed_nodes_in.to(graph_device)
-            frontier = self.sample_frontier(block_id, g, seed_nodes_in)
-
-            # Removing edges from the frontier for link prediction training falls
-            # into the category of frontier postprocessing
-            if exclude_eids is not None:
-                parent_eids = frontier.edata[EID]
-                parent_eids_np = _tensor_or_dict_to_numpy(parent_eids)
-                located_eids = _locate_eids_to_exclude(parent_eids_np, exclude_eids)
-                if not isinstance(located_eids, Mapping):
-                    # (BarclayII) If frontier already has a EID field and located_eids is empty,
-                    # the returned graph will keep EID intact.  Otherwise, EID will change
-                    # to the mapping from the new graph to the old frontier.
-                    # So we need to test if located_eids is empty, and do the remapping ourselves.
-                    if len(located_eids) > 0:
-                        frontier = transform.remove_edges(
-                            frontier, located_eids, store_ids=True)
-                        frontier.edata[EID] = F.gather_row(parent_eids, frontier.edata[EID])
-                else:
-                    # (BarclayII) remove_edges only accepts removing one type of edges,
-                    # so I need to keep track of the edge IDs left one by one.
-                    new_eids = parent_eids.copy()
-                    for k, v in located_eids.items():
-                        if len(v) > 0:
-                            frontier = transform.remove_edges(
-                                frontier, v, etype=k, store_ids=True)
-                            new_eids[k] = F.gather_row(parent_eids[k], frontier.edges[k].data[EID])
-                    frontier.edata[EID] = new_eids
+
+            frontier = self.sample_frontier(block_id, g, seed_nodes_in, exclude_eids=exclude_eids)
```

Comment:
This is moving the edge ID exclusion into `sample_frontier`.  The reason we have edge ID exclusion after `sample_frontier` is that we would like to let user write their own neighbor samplers without taking care of edge exclusion.  Removing it like this will potentially break existing code since the edges are no longer automatically removed and users might do extra work.  @zheng-da do you think it's better to straight move it into `sample_frontier`?

Reply:
After discussion with Minjie we decide it's better to keep the original behavior so existing code will not break.  To have the best of both worlds we can add a flag in `BlockSampler`, say `exclude_edges_in_sample_frontier`.  When it is True, BlockSampler assumes that the edges must be excluded in `sample_frontier`, and `sample_blocks` no longer removes the edges.  Otherwise, it resorts to the current behavior where a frontier is first sampled then with edges removed.  The argument defaults to False, and could be set to True in `MultiLayerNeighborSampler` so `sample_neighbors` with `exclude_edges` can still be utilized.